### PR TITLE
Change how we detect unfilled slots/registers

### PR DESCRIPTION
**Pull request by vjeko**
Rather than relying on a responsive server set, base the decision
on the latest committed layout and the highest WrongEpochException.

## Overview

Description:

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

**Comment by vjeko** [comment]
This is an early preview. This change will be amended with tests.

File: infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java [hunk 1]
```diff
@@ -342,7 +342,7 @@ private synchronized void runDetectionTasks() {
             Layout updatedLayout = correctWrongEpochs(pollReport, layout);
 
             Result<DetectorTask, RuntimeException> failure = Result.of(() -> {
-                if (pollReport.isCurrentLayoutSlotUnFilled()) {
+                if (pollReport.getLayoutSlotUnFilled(updatedLayout).isPresent()) {
                     detectFailure(updatedLayout, Collections.emptySet(), pollReport).join();
```
Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java]: Can we add a log statement here to know that this layout slot was filled with the previous layout and no new cluster reconfigurations have taken place?

Comment (vjeko) [reply infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java]: Ack.

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java [hunk 2]
```diff
@@ -83,23 +86,34 @@
...
+
+    /**
+     * Given the latest committed layout, figure out if the majority of the layout servers have
+     * been sealed.
+     *
+     * @param latestCommittedLayout latest commited layout
+     * @return true f the majority of the layout servers have been sealed.
+     */
+    public boolean isMajorityLayoutServersSealed(Layout latestCommittedLayout) {
+        return wrongEpochs.size() > latestCommittedLayout.getLayoutServers().size()/2;
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java]: code reformat please

File: infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java [hunk 3]
```diff
@@ -339,11 +339,27 @@ private synchronized void runDetectionTasks() {
...
-                if (pollReport.isCurrentLayoutSlotUnFilled()) {
-                    detectFailure(updatedLayout, Collections.emptySet(), pollReport).join();
+
+                // This is just an optimization in case we receive a WrongClusterException
+                // while one of the other management clients is trying to move to a new layout.
+                // This check is merely trying to minimize the the scenario in which we end ip
+                // filling the slot with an outdated layout. Due to potential failures, we cannot
+                // wait for more than the majority of layout servers.
+                if (!pollReport.isMajorityLayoutServersSealed(updatedLayout)) {
+                    log.info(("Majority of the layout servers have not been sealed. Skipping.");
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java]: maybe we can reduce it to trace? In the other case, we would have many many of this messages in the log

File: infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java [hunk 4]
```diff
@@ -339,11 +339,27 @@ private synchronized void runDetectionTasks() {
...
+                    return DetectorTask.COMPLETED;
+                }
+
+                Optional<Long> unfilledSlot = pollReport.getLayoutSlotUnFilled(updatedLayout);
+                // If the latest slot has not been filled, fill it with the previous known layout.
+                if (unfilledSlot.isPresent()) {
+                    log.info("Trying to fill an unfilled slot {}.", unfilledSlot.get());
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java]: looks like we need more info in the log, can you please add `pollReport` into the log

File: infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java [hunk 5]
```diff
@@ -339,11 +339,27 @@ private synchronized void runDetectionTasks() {
...
+                    DetectorTask slotFillStatus = detectFailure(
+                            updatedLayout, Collections.emptySet(), pollReport).join();
+                    log.info("Slot fill was {}.", slotFillStatus);
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java]: do we need it in the log?

**Comment by WenbinZhu** [comment]
> More comments later

Sorry, nvm, I was intended to commented to another patch.

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java [hunk 6]
```diff
@@ -45,6 +47,7 @@
     /**
      * List of responsive servers in a current layout
      */
+    @Deprecated
```
Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java]: Can we remove this if not required?
Is this used in a lot of places?

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java [hunk 7]
```diff
@@ -83,23 +86,32 @@
...
+            return Optional.of(maxOutOfPhaseEpoch);
         }
-        return result;
+
+        return Optional.empty();
+    }
+
+
+    /**
+     * Check tp see if all responsive servers from this {@link PollReport} have been sealed.
```
Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java]: Check to see

File: infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java [hunk 8]
```diff
@@ -339,11 +339,26 @@ private synchronized void runDetectionTasks() {
...
 
             Result<DetectorTask, RuntimeException> failure = Result.of(() -> {
-                if (pollReport.isCurrentLayoutSlotUnFilled()) {
-                    detectFailure(updatedLayout, Collections.emptySet(), pollReport).join();
+
+                // This is just an optimization in case we receive a WrongClusterException
+                // while one of the other management clients is trying to move to a new layout.
+                // This check is merely trying to minimize the the scenario in which we end ip
+                // filling the slot with an outdated layout.
+                if (!pollReport.areAllResponsiveServersSealed()) {
```
Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java]: The check and the log are counter-intuitive. We enter the if statement if all responsive servers are not sealed. However, we log that all servers have been sealed.
Is this logic correct?

Comment (vjeko) [reply infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java]: You should take a look at the latest patch and see if it makes more sense.

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java (hunk 6 above)
Comment (vjeko) [reply infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java]: Yep.

File: infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java [hunk 9]
```diff
@@ -339,11 +339,27 @@ private synchronized void runDetectionTasks() {
...
             // performs re-sealing of all nodes if required and catchup of a layout server to
             // the current state.
-            Layout updatedLayout = correctWrongEpochs(pollReport, layout);
+            final Layout updatedLayout = correctWrongEpochs(pollReport, layout);
 
             Result<DetectorTask, RuntimeException> failure = Result.of(() -> {
-                if (pollReport.isCurrentLayoutSlotUnFilled()) {
-                    detectFailure(updatedLayout, Collections.emptySet(), pollReport).join();
+
+                // This is just an optimization in case we receive a WrongClusterException
```
[the remark this answers is not in the document]
Comment (vjeko) [reply infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java]: Yep.

**Comment by vjeko** [comment]
@zalokhan The best you can do is wait a bit longer for the other management server/client to try to finish sealing and propose a new layout. We cannot indefinitely wait. So the current algorithm is:

On a WrongEpochException (WEE)
1) If total number of WEE is less than the total number of responsive servers, this means we have encountered the scenario in which some other node is in process of proposing a new layout. So we re-seal with the same epoch and back-off, giving an opportunity for the new layout to be accepted.
2) If total number of WEE is equal to the total number of responsive servers, this either means that:
a) We have previously encountered (1) and the other management client is dead or is unable to get the agreement on that register.
b) We have observed a state where the other management client has finished sealing and is in process of proposing. 

As mentioned the comment WRT (a) and (b) -- we cannot eliminate the race, just minimize it.

It is might be possible to make the logic a bit simpler, but this requires adding extra PollReport from the previous iteration.

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java (hunk 2 above)
Comment (vjeko) [reply infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java]: Not sure I understand -- what's wrong with it? Also that code has changed.

File: infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java (hunk 3 above)
Comment (vjeko) [reply infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java]: Reduced it to debug, This message will be printed only during WrongEpochException, so it should not cause much chatter.

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java (hunk 7 above)
Comment (vjeko) [reply infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java]: Done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java (hunk 8 above)
Comment (vjeko) [reply infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java]: I see now. The message is wrong. Fixed.

File: infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java (hunk 4 above)
Comment (vjeko) [reply infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java]: Done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java (hunk 5 above)
Comment (vjeko) [reply infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java]: Removed it.

File: infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java [hunk 10]
```diff
@@ -339,11 +339,27 @@ private synchronized void runDetectionTasks() {
...
+                // while one of the other management clients is trying to move to a new layout.
+                // This check is merely trying to minimize the the scenario in which we end ip
```
[the remark this answers is not in the document]
Comment (vjeko) [reply infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java]: Done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java (hunk 6 above)
Comment (vjeko) [reply infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java]: No longer relevant.

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java [hunk 11]
```diff
@@ -83,23 +86,34 @@
     }
 
     /**
-     * All active Layout servers have been sealed but there is no client to take this forward and
-     * fill the slot by proposing a new layout. This is determined by the outOfPhaseEpochNodes map.
+     * Check to see if the current epoch]register has been decided/filled. We do this by
```
[the remark this answers is not in the document]
Comment (vjeko) [reply infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java]: Fixed.

**Comment by vjeko** [comment]
@zalokhan There is an IT that verifies the functionality.

**Pull request merged**